### PR TITLE
refactor test_sdpa into two test classes to account for failure modes

### PR DESCRIPTION
**Pull request by drisspg**
### Summary
This PR creates a new TestSDPAFailureModes test class in order to better seperate what each test is trying to do.

### CODE REVIEW

**Review by malfet — APPROVED** [review]

File: test/test_transformers.py [hunk 1]
```diff
@@ -66,6 +66,50 @@ def get_rtol(true_value: torch.Tensor, computed_value: torch.Tensor) -> float:
...
+
+backend_map = {
+    SDPBackend.MATH: {"enable_math": True, "enable_flash": False, "enable_mem_efficient": False},
+    SDPBackend.FLASH_ATTENTION: {"enable_math": False, "enable_flash": True, "enable_mem_efficient": False},
+    SDPBackend.EFFICIENT_ATTENTION: {
+        "enable_math": False, "enable_flash": False, "enable_mem_efficient": True}
+}
+
+
+def rand_tensor(shape: Tuple[Union[int, List[int]]], device: str, dtype: torch.dtype, type: str,
```
[the remark this answers is not in the document]
Comment (drisspg) [reply test/test_transformers.py:78]: The name should actually probably be a little different than make_tensor and really be like `rand_sdpa_tensor` since it makes a 4 dim batch_size, num_heads, seq_len, head_dim formatted tensor

File: test/test_transformers.py [hunk 2]
```diff
@@ -997,6 +1041,272 @@ def test_script_mha_in_proj_weight_none(self):
...
+        size = (2, 2, 4, head_dim)
+        q, k, v = make_tensor(size), make_tensor(size), make_tensor(size)
+        with sdp_kernel(enable_mem_efficient=True, enable_flash=False, enable_math=False):
+            self.assertRaises(RuntimeError, lambda: torch.nn.functional.scaled_dot_product_attention(
+                q, k, v, None, 0.0, False))
+
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not isSM86or89Device,
+                     "Does not support fused SDPA or not SM86+ hardware")
+    @parametrize("head_dim", [72, 96, 128])
+    def test_flash_backward_failure_sm86plus(self, head_dim: int):
```
Comment (ngimel) [inline test/test_transformers.py:1067]: same here (turns out nvidia wasn't running on the tests that don't have `cuda` in the name, so they didn't catch a few bugs on the hardware that we don't run CI on, but they do.

File: test/test_transformers.py [hunk 3]
```diff
@@ -997,6 +1041,272 @@ def test_script_mha_in_proj_weight_none(self):
...
+    def test_invalid_fused_inputs_dim_3(self, kernel: SDPBackend):
+        with sdp_kernel(**backend_map[kernel]):
+            # Dim is not 4
+            device = "cuda"
+            size = (2, 3, 8)
+            dtype = torch.float16
+            q = torch.randn(size, device=device, dtype=dtype)
+            k = torch.randn(size, device=device, dtype=dtype)
+            v = torch.randn(size, device=device, dtype=dtype)
+            self.assertRaises(RuntimeError, lambda: torch.nn.functional.scaled_dot_product_attention(
```
Comment (ngimel) [inline test/test_transformers.py:1121]: do you want RaisesRegexp here?

File: test/test_transformers.py [hunk 4]
```diff
@@ -997,6 +1041,272 @@ def test_script_mha_in_proj_weight_none(self):
...
+
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not SM80OrLater, "Does not support fused SDPA or pre-SM80 hardware")
+    def test_flash_fail_fp32(self):
+        device = 'cuda'
+        dtype = torch.float
+        shape = (16, 16, 32, 32)
+        make_tensor = partial(rand_tensor, shape=shape, type=type, device=device, dtype=dtype)
+        q, k, v = make_tensor(), make_tensor(), make_tensor()
+        with sdp_kernel(enable_flash=True, enable_mem_efficient=False, enable_math=False):
+            self.assertRaises(RuntimeError, lambda: torch.nn.functional.scaled_dot_product_attention(
```
Comment (ngimel) [inline test/test_transformers.py:1212]: RaisesRegexp would make sense

File: test/test_transformers.py [hunk 5]
```diff
@@ -997,6 +1041,272 @@ def test_script_mha_in_proj_weight_none(self):
...
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not SM80OrLater, "Does not support SDPA or pre-SM80 hardware")
+    def test_flash_autocast_fp32_bfloat16(self):
+        device = 'cuda'
+        dtype = torch.float
+        shape = (16, 16, 32, 32)
+        make_tensor = partial(rand_tensor, shape=shape, type=type, device=device, dtype=dtype)
+        q, k, v = make_tensor(), make_tensor(), make_tensor()
+        with torch.autocast(device_type=device, dtype=torch.bfloat16):
+            with sdp_kernel(enable_flash=True, enable_mem_efficient=False, enable_math=False):
+                _ = torch.nn.functional.scaled_dot_product_attention(
```
Comment (ngimel) [inline test/test_transformers.py:1236]: this looks like something that should work on cpu also, no? And on prev hardware versions?

File: test/test_transformers.py [hunk 6]
```diff
@@ -997,6 +1041,272 @@ def test_script_mha_in_proj_weight_none(self):
...
+        kpm = ones_tensor(S, L) * float("-inf")
+        am = ones_tensor(L, L).to(bool)
+
+        def func():
+            return mha(qkv, qkv, qkv, need_weights=False, key_padding_mask=kpm, attn_mask=am)
+
+        self.assertRaises(RuntimeError, func)
+
+    @parametrize("kernel", [SDPBackend.MATH, SDPBackend.FLASH_ATTENTION, SDPBackend.EFFICIENT_ATTENTION])
+    @parametrize("device", ["cpu", "cuda"] if TEST_CUDA else ["cpu"])
```
Comment (ngimel) [inline test/test_transformers.py:1258]: do you set `device_list` somewhere to not repeat this `if TEST_CUDA`?

File: test/test_transformers.py [hunk 7]
```diff
@@ -997,6 +1041,272 @@ def test_script_mha_in_proj_weight_none(self):
...
+class TestSDPAFailureModes(NNTestCase):
+    """ Used to test the failure modes of scaled_dot_product_attention
+    """
+    _do_cuda_memory_leak_check = True
+    _do_cuda_non_default_stream = True
+
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_SDPA or not isSM86or89Device,
+                     "Does not support fused SDPA or not SM86+ hardware")
+    @parametrize("head_dim", [72, 96, 128])
+    def test_memory_efficient_sm86_plus_failure(self, head_dim: int):
```
[the remark this answers is not in the document]
Comment (malfet) [reply test/test_transformers.py:1053]: Can we just have a `device` argument and decorate with `@onlyCUDA`?

**Comment by ngimel** [comment]
@drisspg what would be needed to address TODO at the bottom and instantiate device type tests? That would automatically enforce device arg everywhere
https://github.com/drisspg/pytorch/blob/8e6c527039b8f6b185f6d4fcb48ca8f50ee7c3b9/test/test_transformers.py#L2071-L2075

File: test/test_transformers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (drisspg) [reply test/test_transformers.py]: I did a largeish refactor to fix the TODO and use instantiate_device_type_tests. For this test I used @onlyCUDA. I don't need to do the name changes right? Since instantiate_device_type_tests will stub out the _cuda version test?

File: test/test_transformers.py (hunk 5 above)
Comment (drisspg) [reply test/test_transformers.py:1239]: Yeah this should work but is tested in common_method_invocations. This was more just to specifcally test that flash on newer hardward works since this was added manually

**Comment by drisspg** [comment]
@ngimel @malfet I made some substantial changes to the device handiling to address the todo and addressed the regex comments.

**Review by ngimel — APPROVED** [review]
awesome, thanks!
